diff --git a/drivers/mtd/nand/brcmstb_nand.c b/drivers/mtd/nand/brcmstb_nand.c
index 97bd867..c28e090 100644
--- a/drivers/mtd/nand/brcmstb_nand.c
+++ b/drivers/mtd/nand/brcmstb_nand.c
@@ -1268,6 +1268,11 @@ static int __devinit brcmstb_nand_probe(struct platform_device *pdev)
 		goto out;
 	}
 
+	//Timing Adjust
+        BDEV_WR_RB(REG_TIMING_1(host->cs), 0x4232222D);
+        BDEV_WR_RB(REG_TIMING_2(host->cs), 0x1a84);
+
+
 #ifdef CONFIG_MTD_PARTITIONS
 	nr_parts = parse_mtd_partitions(mtd, part_probe_types, &parts, 0);
 	if (nr_parts <= 0) {
